File: python/perfect-numbers/perfect_numbers.py
import operator

classifications = [(operator.lt, 'deficient'),
                   (operator.eq, 'perfect'),
                   (operator.gt, 'abundant')]

def classify(number):
    if number <= 0:
        raise ValueError('Classification is only possible for positive integers.')

    aliquot_sum = sum(n for n in range(1, number) if number % n == 0)

    return next(result for operation, result in classifications if operation(aliquot_sum, number))

File: javascript/perfect-numbers/perfect-numbers.js
function * divisors(num) {
    const max = Math.floor(num/2);
    for (let x = 1; x <= max; x++) {
        if (num % x === 0) {
            yield x;
        }
    }
}

export function classify(num) {
    if (num < 1) {
        throw new Error('Classification is only possible for natural numbers.');
    }

    const sum = [...divisors(num)].reduce((a, b) => a + b, 0);

    return sum < num ? 'deficient'
         : sum > num ? 'abundant'
         : 'perfect';
}
